NickAkhmetov/React 19 3D/XR support (#2415)
vitessce/vitessce: sites/docs/docs/upgrade-guide.md

This guide describes how to upgrade from one major version of the `vitessce` JavaScript package to the next.

## v3 to v4

### React 19

Vitessce v4 is built against React 19.

- React 18.3+ is required for 2D views.
- React 19 is required for 3D / spatial / VR (XR) views, because Vitessce now uses
  `@react-three/fiber` v9, which only supports React 19.
- React 16 and 17 are no longer supported.

If your app still uses the legacy `ReactDOM.render`, switch to `createRoot`:

```diff
- import ReactDOM from 'react-dom';
- ReactDOM.render(<Vitessce {...props} />, document.getElementById('root'));
+ import { createRoot } from 'react-dom/client';
+ createRoot(document.getElementById('root')).render(<Vitessce {...props} />);
```

> 

Reviewer: ```suggestion

For further React upgrade information, see the [React 19 Upgrade Guide](https://react.dev/blog/2024/04/25/react-19-upgrade-guide).
```

### Three.js and React Three Fiber peer dependencies

> 

Reviewer: ```suggestion

The following information is only relevant for Vitessce users who use the ThreeJS-based `spatial-three` or `spatial-accelerated` views for 3D volumetric and mesh rendering (which includes users of [Vitessce Link](https://vitessce.link/about) and its VR rendering features). Users who only rely on the DeckGL- and Neuroglancer-based 3D rendering functionality in Vitessce can ignore this section.

```

I think we should state who the section applies to up-front.

Previously, `three` and the `@react-three/*` packages were bundled into Vitessce.
In v4 they are **peer dependencies**, so that a single copy is shared with your
application. (Multiple copies of `@react-three/fiber` cause runtime errors such as
`"R3F: Hooks can only be used within the Canvas component!"`.)

> If you use any 3D, spatial, volume-rendering, or VR views, install them alongside

Reviewer: ```suggestion
If you use the ThreeJS-based functionality, install the required dependencies alongside
```

Vitessce:

```sh
npm install three @react-three/fiber @react-three/drei
# Only if you use VR / XR views:
npm install @react-three/xr
```

Recommended versions (with React 19):

| Package | Version |
| --- | --- |
| `three` | `>=0.159.0` |
| `@react-three/fiber` | `^9.0.0` |
| `@react-three/drei` | `^10.0.0` |
| `@react-three/xr` | `^6.0.0` |

If you already had `three` / `@react-three/fiber` installed, bump them to these

> versions. Note that `@react-three/fiber` v9 requires React 19.

Reviewer: ```suggestion
versions. Note that `@react-three/fiber` v9 requires React 19. For more details, visit the [react-three-fiber documentation](https://r3f.docs.pmnd.rs/getting-started/introduction).
```

This guide describes how to upgrade from one major version of the `vitessce` JavaScript package to the next.

## v3 to v4

### React 19

> Vitessce v4 is built against React 19.

Reviewer: ```suggestion
Vitessce v4 is now developed and tested against React 19.
```

- React 18.3+ is required for 2D views.
- React 19 is required for 3D / spatial / VR (XR) views, because Vitessce now uses

>   `@react-three/fiber` v9, which only supports React 19.

Reviewer: ```suggestion
- Most functionality remains compatible with React 18.3+.
- React 19 is required for usage of the ThreeJS-based 3D rendering functionality, because this uses
  `@react-three/fiber` v9, which only supports React 19. See the instructions below for more details.
```
